blog/_posts/2023-12-08-migration-to-scala-3.md

```diff
@@ -0,0 +1,153 @@
+
+It’s also an excellent time to discuss the benefits and trade-offs. And keep it fair. For example, if you just want a new syntax, is it worth disturbing an old service that hasn’t been touched in years? Well, it’s up to you to decide.
+
+*We decided to write all the new services in Scala 3 (where most of the future work will happen) and keep the rest. Consequently, onboarding new devs without prior Scala experience is one of my biggest challenges. When they switch from Scala 3 service to Scala 2, they bump into hurdles — some of the “intuitive” concepts don’t work as expected anymore. This requires more explanation; they seem to have to learn some things twice.*
+
+Okay, we defined what migration means, but how exactly do we go about it?
+
+## Step 1. Analyze and plan
+
+Even if we decide not to touch the existing services, we need to review our codebase:
+
+- Do we have any shared internal libraries? Would we need to use them in new services?
+- What external libraries do we use? Do they have a Scala3-compatible release?
+
+Other questions you might need to consider:
+
+- How many services do you need to migrate?
+- What Scala version(s) does each service use?
+
+Thanks to the [interoperability](https://docs.scala-lang.org/scala3/guides/migration/compatibility-source.html) between Scala 3 and Scala 2.13, it might be more accessible or crucial to upgrade to Scala 2.13 first. You have to keep this in mind.
+
+*By this point, we knew that all our services were on 2.13; we knew how many internal libraries needed to be used from both Scala versions and what external libraries had no Scala 3 support: what the status is, what is missing, and what we can do to deal with it. For example, a couple of libraries got a Scala 3 release by the time we were done with the investigation, and one `circe` ”extras” library was replaced with two lines of boilerplate.*
+
+## Step 2. Draw the rest of the owl
+
+After all the meetings and/or writings comes the time to do the actual work. When you have concrete questions or action items, remember that [Scala 3 migration guides](https://docs.scala-lang.org/scala3/guides/migration/compatibility-intro.html) are your friends. For example, they cover how to port compiler options and build configurations. Great reference resource.
+
+> Note that you can find further help from the community on [Scala Discord](https://discord.com/invite/scala).
+
+We don’t have to go into these details right now. Instead, let’s talk about the bigger picture and a few other things to remember.
+
+### Dealing with internal libraries
+
+There are multiple ways of sharing a library between Scala 2 and Scala 3 services.
+
+**The good news** is that you can use Scala 2.13 libraries from Scala 3 app and Scala 3 libraries from Scala 2.13 app.
+
+**The bad news** is that you can’t do it directly if you have shared transitive dependencies.
+
+**The good news** is that it’s pretty easy to deal with.
```

Comment:
"good" then "bad" then "good" doesn't work as a "flow". Consider using dedicated connectors such as "However" and "Therefore" instead of repeating the same "news" pattern.